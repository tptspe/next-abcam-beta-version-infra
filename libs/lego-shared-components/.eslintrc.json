{
  "extends": [
    "plugin:@nrwl/nx/react",
    "../../.eslintrc.json",
    "plugin:jsx-a11y/recommended",
    "plugin:cypress/recommended"
  ],
  "ignorePatterns": ["!**/*", "*.stories.*", "*.spec.*", ".storybook/*"],
  "plugins": ["no-relative-import-paths"],
  "rules": {
    "no-relative-import-paths/no-relative-import-paths": [
      "warn",
      { "allowSameFolder": true }
    ]
  },
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "parserOptions": {
        "project": ["libs/lego-shared-components/tsconfig.*?.json"]
      },
      "rules": {}
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    }
  ]
}
