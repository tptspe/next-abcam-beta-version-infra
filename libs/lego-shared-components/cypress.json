{
  "fileServerFolder": ".",
  "fixturesFolder": "./e2e/fixtures",
  "integrationFolder": "./e2e/integration",
  "modifyObstructiveCode": false,
  "pluginsFile": "./e2e/plugins/index",
  "supportFile": "./e2e/support/index.ts",
  "video": true,
  "videosFolder": "../../dist/cypress/libs/lego-shared-components/videos",
  "screenshotsFolder": "./e2e/screenshots",
  "screenshotOnRunFailure": false,
  "videoCompression": false,
  "chromeWebSecurity": false
}
