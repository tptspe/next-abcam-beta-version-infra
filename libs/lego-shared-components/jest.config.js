module.exports = {
  displayName: 'lego-shared-components',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  moduleNameMapper: {
    '\\.svg$': '<rootDir>/mocks/svg-loader.js',
  },
  coverageDirectory: '../../coverage/libs/lego-shared-components',
}
