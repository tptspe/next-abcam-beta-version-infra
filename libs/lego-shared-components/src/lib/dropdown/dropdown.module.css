.dropdown {
  @apply text-body-small;
  @apply font-sb;
  @apply relative;
  @apply inline-flex;
  @apply flex-col;
  @apply cursor-pointer;

  .selectedValueContainer {
    @apply bg-interactiveWhiteTransparentActive;
    @apply text-white;
    @apply flex;
    @apply flex-row;
    @apply gap-2;
    @apply py-2.5;
    @apply px-4;
    @apply rounded-sm;
    @apply border border-solid border-strokeDarkBGMedium;
    width: fit-content;

    &:focus {
      @apply outline-none;
      @apply shadow-interactiveElement;
    }

    .openIcon {
      @apply w-5;
      @apply h-5;
      @apply border-strokeDarkBGMedium;
      path {
        @apply fill-current;
      }
    }

    &:hover {
      @apply bg-strokeDarkBGMedium;
    }
  }

  .dropdownContent {
    @apply rounded-lg;
    margin-top: 0.375rem;
  }
}

.secondary {
  .selectedValueContainer {
    @apply bg-interactiveGreyTransparentHover;
    @apply border-interactiveGreyTransparentHover;
    @apply text-grey20;

    &:hover {
      @apply shadow-interactiveElement;
      @apply bg-interactiveGreyTransparentHover;
    }
    &:focus {
      @apply outline-none;
      @apply shadow-interactiveElement;
    }
  }
}
