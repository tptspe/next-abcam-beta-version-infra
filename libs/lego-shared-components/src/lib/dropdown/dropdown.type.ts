import { ReactElement } from 'react'

export type DropdownOption = {
  key: string
  displayValue: string
  shortDisplayValue?: string
}

export enum DropdownVariant {
  primary = 'primary',
  secondary = 'secondary',
}

export type DropdownProps = {
  options: DropdownOption[]
  onChange: (val: DropdownOption) => void
  children: ReactElement
  variant?: DropdownVariant
}
