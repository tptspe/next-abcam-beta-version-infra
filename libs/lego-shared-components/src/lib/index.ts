export { Button } from './button/button'
export { Select } from './select/select'
export { Dropdown } from './dropdown/dropdown'
