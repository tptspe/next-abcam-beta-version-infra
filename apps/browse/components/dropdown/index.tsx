import React, { createContext } from 'react'
import styled, { css } from 'styled-components'
import { Button } from '@browse/components/button'
import { ReactComponent as ChevronUp } from '@browse/public/icons/chevron-up.svg'
import { ReactComponent as Cross } from '@browse/public/icons/cross.svg'
import { Checkbox } from '@browse/components/checkbox'
import { green35, white } from '@browse/public'
import { useOnClickOutside } from '@browse/hooks/useOnClickOutside'
import { ChevronDown } from '@abcam-web/lego-shared-components/icons'

interface WrapperProps {
  show: boolean
  size: 'small' | 'medium'
}

const Wrapper = styled.div<WrapperProps>`
  position: absolute;
  top: 3rem;
  ${({ size }) =>
    size === 'small' &&
    `
    top: 2rem;
  `}
  min-width: 14rem;
  z-index: 999;
  border-radius: 8px;
  box-shadow: 0 16px 32px 0 rgba(0, 0, 0, 0.2);
  background-color: ${white};
  display: none;
  ${({ show }) =>
    show &&
    `
    display: flex;
    flex-direction: column;
    padding-right: 1rem;
  `}
`

const FilterButton = styled(Button)`
  &:focus {
    border: 3px solid ${green35} !important;
    overflow: hidden;
    margin: -3px;
  }
`

const ActionButtonWrapper = styled.div`
  display: flex;
  padding: 0.5rem 1rem 1rem 1rem;
  justify-content: space-between;
`

const ActionButton = styled(Button)`
  ${({ theme }) => css`
    border: solid 1px ${theme.color.lightTertiaryButtonDisabledBorderColor} !important;
  `}
`
interface WrapperButtonProps {
  style?: any
}

const WrapperButton = styled.div<WrapperButtonProps>`
    position: relative;
    ${({ style }) => style}
}
`

const Close = styled(Cross)`
  cursor: pointer;
`

const ChevronDownIcon = styled(ChevronDown)`
  ${({ theme }) =>
    `path {
        fill: ${theme.color.textGreyDark};
    }`}
`

const ChevronUpIcon = styled(ChevronUp)`
  ${({ theme }) =>
    `path {
        fill: ${theme.color.textGreyDark};
    }`}
`

export enum DropDownType {
  Single = 'SINGLE',
  Multi = 'MULTI',
}

export enum SelectorType {
  CheckBox = 'checkbox',
  RadioButton = 'radio',
  Base = 'base',
}

type Sizes = 'small' | 'medium'

export interface DropdownProps {
  label?: React.ReactElement
  type: DropDownType
  selectorType: SelectorType
  onSelect?: (selected: string) => void
  size?: Sizes
  style?: any
  clearAction?: (label: string) => void
  showFooter?: boolean
}

const DropDownContext = createContext({
  type: DropDownType.Single,
  selectorType: SelectorType.CheckBox,
  checkedList: [] as string[],
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  toggleCheckDropDown: (value: string, checked: boolean) => {},
})

export const Dropdown: React.FC<DropdownProps> = ({
  children,
  selectorType = SelectorType.Base,
  showFooter = false,
  size = 'medium',
  ...props
}) => {
  const [show, setShow] = React.useState(false)
  const [selected, setSelected] = React.useState([] as string[])
  const ref = useOnClickOutside(() => setShow(false))
  return (
    <WrapperButton data-cy={'drop-down'} ref={ref}>
      <FilterButton
        variant={'tertiary'}
        rightIcon={show ? <ChevronUpIcon /> : <ChevronDownIcon />}
        size={size}
        onClick={() => setShow((show) => !show)}
      >
        {props.label}
      </FilterButton>

      <Wrapper show={show} size={size}>
        <DropDownContext.Provider
          value={{
            type: props.type,
            selectorType,
            checkedList: selected,
            toggleCheckDropDown: (value: string, checked: boolean) => {
              if (checked) {
                if (props.type === DropDownType.Multi) {
                  setSelected(selected.concat(value))
                } else {
                  setSelected([value])
                }

                return
              }

              const selectedLocal = Object.assign([], selected)
              const removeItemIndex = selected.indexOf(value, 0)

              if (removeItemIndex > -1) {
                selectedLocal.splice(removeItemIndex, 1)
                setSelected(selectedLocal)
              }
              return
            },
          }}
        >
          <Item>{children}</Item>
        </DropDownContext.Provider>

        {showFooter && (
          <ActionButtonWrapper>
            {/*<ActionButton variant='tertiary' size='small' onClick={props.clearAction}>*/}
            {/*    Clear*/}
            {/*</ActionButton>*/}

            <ActionButton
              data-cy={'close-button'}
              variant="tertiary"
              size="small"
              leftIcon={<Close />}
              onClick={() => setShow(false)}
            >
              Close
            </ActionButton>
          </ActionButtonWrapper>
        )}
      </Wrapper>
    </WrapperButton>
  )
}

const Item = styled.div`
  display: flex;
  flex-direction: column;
`

export const DropDownRow = styled.div`
  ${({ theme }) => css`
    padding: 1rem ${theme.size.layout[1]};
  `};

  cursor: pointer;

  :hover {
    background-color: rgba(39, 63, 63, 0.05);
  }

  font-size: 0.75rem;
  font-weight: 'normal';
  font-stretch: normal;
  font-style: normal;
  letter-spacing: 0.5px;
`

interface DropDownOptionProps {
  index: number
  label: string
  count: number
  value: string
}

export const DropDownOption: React.FC<DropDownOptionProps> = (props) => {
  return (
    <DropDownContext.Consumer>
      {(context) => (
        <>
          {(context.selectorType === SelectorType.CheckBox ||
            context.selectorType === SelectorType.RadioButton) && (
            <DropDownRow>
              <Checkbox
                label={props.label}
                type={context.selectorType}
                name={`${context.selectorType}-${props.index}`}
                id={`${context.selectorType}-${props.index}`}
                checked={context.checkedList.includes(props.value)}
                onCheckChange={(isSelected) => {
                  context.toggleCheckDropDown(props.value, !!isSelected)
                }}
              />
            </DropDownRow>
          )}

          {context.selectorType === SelectorType.Base && (
            <DropDownRow
              onClick={() => {
                context.toggleCheckDropDown(props.value, true)
              }}
            >
              <span>{props.label}</span>
            </DropDownRow>
          )}
        </>
      )}
    </DropDownContext.Consumer>
  )
}
