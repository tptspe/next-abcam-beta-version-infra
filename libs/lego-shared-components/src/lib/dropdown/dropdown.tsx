import React, { cloneElement, useCallback, useRef, useState, FC } from 'react'
import styles from './dropdown.module.css'
import { ChevronDown, ChevronUp } from '@abcam-web/lego-shared-components/icons'
import { useClickOutside } from '@abcam-web/lego-shared-components/lib/utils/useClickOutside'
import classnames from 'classnames'
import {
  DropdownProps,
  DropdownVariant,
} from '@abcam-web/lego-shared-components/lib/dropdown/dropdown.type'
import {
  isArrowDownEvent,
  isEnterEvent,
  isSpaceEvent,
} from '@abcam-web/lego-shared-components/lib/utils/keyboardEvents'

const Dropdown: FC<DropdownProps> = (props) => {
  const containerRef = useRef(null)

  const {
    options,
    children,
    onChange,
    variant = DropdownVariant.primary,
  } = props

  const [initialValue] = options
  const [selectedValue, setSelectedValue] = useState(initialValue)
  const [isOpen, setIsOpen] = useState(false)

  const toggleIsVisible = useCallback(() => {
    setIsOpen(!isOpen)
  }, [isOpen, setIsOpen])

  const hideOnOutsideClick = useCallback(() => {
    if (isOpen) {
      setIsOpen(false)
    }
  }, [isOpen, setIsOpen])

  useClickOutside(containerRef, hideOnOutsideClick)

  const classes = classnames({
    [styles.dropdown]: true,
    [styles.open]: isOpen,
    [styles.secondary]: variant === DropdownVariant.secondary,
  })

  const setSelectedValueCallback = useCallback(
    (value) => {
      setSelectedValue(value)
      toggleIsVisible()
      onChange(value)
    },
    [setSelectedValue, toggleIsVisible, onChange]
  )

  const listKeyUpHandler = useCallback(
    (evt) => {
      if (isEnterEvent(evt) || isSpaceEvent(evt) || isArrowDownEvent(evt)) {
        toggleIsVisible()
      }
    },
    [toggleIsVisible]
  )

  const childWithOptions = cloneElement(children, {
    options,
    onValueSelect: setSelectedValueCallback,
  })

  return (
    <div className={classes} ref={containerRef}>
      <div
        tabIndex={0}
        className={styles.selectedValueContainer}
        role="button"
        onClick={toggleIsVisible}
        onKeyUp={listKeyUpHandler}
      >
        {selectedValue.shortDisplayValue || selectedValue.displayValue}
        {!isOpen && (
          <ChevronDown className={styles.openIcon} aria-label="Open dropdown" />
        )}
        {isOpen && (
          <ChevronUp className={styles.openIcon} aria-label="Open dropdown" />
        )}
      </div>
      {isOpen && (
        <div className={styles.dropdownContent}>{childWithOptions}</div>
      )}
    </div>
  )
}

export { Dropdown }
