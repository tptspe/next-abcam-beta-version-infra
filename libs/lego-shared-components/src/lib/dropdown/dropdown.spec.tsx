import React from 'react'
import { render, screen } from '@testing-library/react'
import { Dropdown, Select } from '@abcam-web/lego-shared-components/lib'
import {
  mockDropdownData,
  mockDropdownDataWithShortDisplayValue,
} from '@abcam-web/lego-shared-components/lib/dropdown/stories/mocks'
import { DropdownVariant } from '@abcam-web/lego-shared-components/lib/dropdown/dropdown.type'
import userEvent from '@testing-library/user-event'

describe('The dropdown element', () => {
  describe('Snapshots', () => {
    it('Matches the snapshot when the data does not have a short display value', () => {
      const { container } = render(
        <Dropdown onChange={() => true} options={mockDropdownData}>
          <span>test</span>
        </Dropdown>
      )

      expect(container).toMatchSnapshot()
    })
    it('Matches the snapshot when the data has a short display value', () => {
      const { container } = render(
        <Dropdown
          onChange={() => true}
          options={mockDropdownDataWithShortDisplayValue}
        >
          <span>test</span>
        </Dropdown>
      )

      expect(container).toMatchSnapshot()
    })

    it('Matches the snapshot when the secondary variant is specified', () => {
      const { container } = render(
        <Dropdown
          onChange={() => true}
          options={mockDropdownDataWithShortDisplayValue}
          variant={DropdownVariant.secondary}
        >
          <span>test</span>
        </Dropdown>
      )

      expect(container).toMatchSnapshot()
    })
  })

  describe('Before the dropdown is clicked', () => {
    it('Does not display its child component', () => {
      render(
        <Dropdown
          onChange={() => true}
          options={mockDropdownDataWithShortDisplayValue}
          variant={DropdownVariant.secondary}
        >
          <input aria-label={'filter-list'} />
        </Dropdown>
      )

      const inputItem = screen.queryAllByLabelText('filter-list')
      expect(inputItem).toEqual([])
    })
  })

  describe('When the dropdown is clicked', () => {
    it('Displays its child component', () => {
      render(
        <Dropdown
          onChange={() => true}
          options={mockDropdownDataWithShortDisplayValue}
          variant={DropdownVariant.secondary}
        >
          <input aria-label={'filter-list'} />
        </Dropdown>
      )
      const clickElement = screen.getByText(
        mockDropdownDataWithShortDisplayValue[0].shortDisplayValue
      )
      userEvent.click(clickElement)

      const inputElement = screen.getByLabelText('filter-list')
      expect(inputElement).not.toBeNull()
    })
  })
})
