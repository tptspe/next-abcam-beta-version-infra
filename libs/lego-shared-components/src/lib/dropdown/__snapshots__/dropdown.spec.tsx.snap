// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The dropdown element Snapshots Matches the snapshot when the data does not have a short display value 1`] = `
<div>
  <div
    class="dropdown"
  >
    <div
      class="selectedValueContainer"
      role="button"
      tabindex="0"
    >
      United Kingdom
      <svg
        aria-label="Open dropdown"
        class="openIcon"
      />
    </div>
  </div>
</div>
`;

exports[`The dropdown element Snapshots Matches the snapshot when the data has a short display value 1`] = `
<div>
  <div
    class="dropdown"
  >
    <div
      class="selectedValueContainer"
      role="button"
      tabindex="0"
    >
      ðŸ‡¬ðŸ‡§
      <svg
        aria-label="Open dropdown"
        class="openIcon"
      />
    </div>
  </div>
</div>
`;

exports[`The dropdown element Snapshots Matches the snapshot when the secondary variant is specified 1`] = `
<div>
  <div
    class="dropdown secondary"
  >
    <div
      class="selectedValueContainer"
      role="button"
      tabindex="0"
    >
      ðŸ‡¬ðŸ‡§
      <svg
        aria-label="Open dropdown"
        class="openIcon"
      />
    </div>
  </div>
</div>
`;
